// grunt-bower/Gruntfile.js
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    nodeunit: {
      files: ['test/**/*.js']
    },
    jshint: {
      files: ['grunt.js', 'tasks/**/*.js', 'test/**/*.js'],
      options: {
        curly: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        newcap: true,
        noarg: true,
        sub: true,
        undef: true,
        boss: true,
        eqnull: true,
        node: true,
        es5: true,
        strict: false,
        laxcomma: true
      },
      globals: {}
    },
    watch: {
      files: '<config:jshint.files>',
      tasks: 'default'
    },
    bower: {
      dev: {
        src: 'foo/',
        dest: 'public/js/lib',
        options: {
          basePath: '/'
        }
      }
    },
    clean: ['public/']
  });

  // Load local tasks.
  grunt.loadTasks('tasks');

  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');

  // Default task.
  grunt.registerTask('default', ['jshint', 'nodeunit', 'bower']);

};
